### PR TITLE
[GEOS-10926] Community Module Proxy-Base-Ext

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOS-10926](https://osgeo-org.atlassian.net/browse/GEOS-10926)

link and graphic

dao first pass

started on URLMangler

mangler first pass

first test

more tests

more tests2

integration tests added

added first bit of REST controller

more rest controller

Created rule bean builder and updated dao and controller

added position to the rule

added position sort

started on rest unit tests

rest test get put

rest post test

transformer input not working

added back position and fixed issue with model calling builder

ui tests

added copyright

cleaned up test resource paths

javadoc

web page description improvement

REST API YAML

first pass at documentation

added test interface and updated documentation

updated usage

spotless

more cleanup

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [x] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/WebLinkingUrlMangler.java [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+     *
+     * @param urlString the URL to transform
+     * @param headersAsProperties the headers as a properties file style string
+     * @return the transformed URL
+     * @throws IOException if the URL is malformed
+     */
+    public String transformURL(String urlString, String headersAsProperties) throws IOException {
+        URL url = new URL(urlString);
+        StringBuilder baseURL =
+                new StringBuilder(url.getProtocol() + PROTOCOL_SEPARATOR + url.getHost());
```
[the remark this answers is not in the document]
Comment (aaime) [reply src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/WebLinkingUrlMangler.java:97]: Actually... according to Jacoco's code coverage, this method is not exercised.... but I see the WEB UI actually calling it. So I'm guessing there are no tests in the web page for it.

File: src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingConfigPage.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+import org.geoserver.proxybase.ext.WebLinkingUrlMangler;
+import org.geoserver.proxybase.ext.config.LinkingManglerRule;
+import org.geoserver.web.GeoServerSecuredPage;
+import org.geoserver.web.wicket.GeoServerDataProvider;
+import org.geoserver.web.wicket.GeoServerTablePanel;
+import org.geoserver.web.wicket.ImageAjaxLink;
+import org.geoserver.web.wicket.ParamResourceModel;
+
+/** The configuration page for the {@link WebLinkingUrlMangler}. */
+public class WebLinkingManglingConfigPage extends GeoServerSecuredPage {
```
Comment (aaime) [inline src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingConfigPage.java:33]: e.g.:

```suggestion
public class ProxyBaseExtConfigPage extends GeoServerSecuredPage {
```

File: src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingRulePage.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    public class SimpleRulePanel extends Panel {
+        /**
+         * Constructor.
+         *
+         * @param panelId the panel id
+         * @param model the model
+         */
+        public SimpleRulePanel(String panelId, IModel<RuleModel> model) {
+            super(panelId, model);
+            add(new NumberTextField<Integer>("position").setMinimum(1).setRequired(true));
```
Comment (aaime) [inline src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingRulePage.java:95]: When creating a new rule the default is "zero":

image

Actually not sure why the minimum has to be 1, but either make the minimum zero, or make sure the default value is sensible.

**Comment by turingtestfail** [comment]
> Testing it interactively, I cannot make it work:
> 
> image
> 
> Why should I select rules? The testing should default on using them all... But even if I select one, it won't work (should, the URL is a match for the regex in the rule...)

Changed to use all rules.  It wasn't working because the protocol is required.  Changed to report URL issues into the output box.

File: doc/en/user/source/community/proxy-base-ext/install.rst [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+#. Visit the :website:`website download <download>` page, locate your release, and download:  :download_extension:`proxy-base-ext`
+   
+   Verify that the version number in the filename (for example |release| above) corresponds to the version of GeoServer you are running.
+
+#. Extract the contents of the archive into the :file:`WEB-INF/lib` directory in GeoServer.
+   Make sure you do not create any sub-directories during the extraction process.
+
+#. Restart GeoServer.
+
+If installation was successful, you will see a new Proxy Base Extension entry in the left menu, under "Settings".
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply doc/en/user/source/community/proxy-base-ext/install.rst]: done

File: doc/en/user/source/community/proxy-base-ext/usage.rst [hunk 5]
```diff
@@ -0,0 +1,66 @@
+.. _proxy_base_extension_usage:
+
+Using the Proxy Base Extension module
+=====================================
+
+This extension allows you to replace URLs in the response of a web service request with a different URL.
+This is useful when you want to use a proxy to redirect your web service requests, but you want to keep the original URL in the response.
+An example of this is when you want to proxy a WMS request to a different server, but you want to keep the original URL in the elements of the GetCapabilities response.
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply doc/en/user/source/community/proxy-base-ext/usage.rst]: done

File: doc/en/user/source/community/proxy-base-ext/usage.rst [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+    https://wfs.eastern.com/mycollection/myfeature
+
+
+Rules Management
+-----------------------------
+
+Rules can be managed and tested with simulated headers in the rules management UI. Besides the basic operations like add, remove and update is also possible to activate or deactivate rules. A deactivated rule will be ignored by this module.
+
+.. figure:: images/proxy_base_rule_config.png
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply doc/en/user/source/community/proxy-base-ext/usage.rst:59]: done

File: doc/en/user/source/community/proxy-base-ext/usage.rst [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+   :align: center
+
+   *Rules management UI*
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply doc/en/user/source/community/proxy-base-ext/usage.rst:62]: done

File: src/community/proxy-base-ext/pom.xml [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+  <parent>
+    <groupId>org.geoserver</groupId>
+    <artifactId>community</artifactId>
+    <version>2.24-SNAPSHOT</version>
+  </parent>
+
+  <groupId>org.geoserver.community</groupId>
+  <artifactId>gs-proxy-base-ext</artifactId>
+  <name>Configurable altering of GeoServer provided web links</name>
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/proxy-base-ext/pom.xml]: done

File: src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/WebLinkingUrlMangler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/WebLinkingUrlMangler.java]: done

File: src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/WebLinkingUrlMangler.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+        try {
+            URL url = new URL(transformerReplacedLiterals);
+            baseURL.setLength(0);
+            baseURL.append(url.getProtocol() + PROTOCOL_SEPARATOR + url.getHost());
+            path.setLength(0);
+            path.append(url.getPath());
+        } catch (MalformedURLException e) {
+            LOGGER.log(
+                    Level.ALL,
+                    "Web Linking URL Mangler: The transformer, after header template replacement: "
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/WebLinkingUrlMangler.java]: done

File: src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/WebLinkingUrlMangler.java [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+    private Map<String, String> collectHeaders(List<String> requiredHeaders) {
+        return requiredHeaders.stream()
+                .filter(headerName -> HTTPHeadersCollector.getHeader(headerName) != null)
+                .collect(Collectors.toMap(item -> item, HTTPHeadersCollector::getHeader));
+    }
+
+    private Map<String, String> collectHeaders(
+            List<String> requiredHeaders, String headersAsProperties) throws IOException {
+        Properties properties = parsePropertiesString(headersAsProperties);
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/WebLinkingUrlMangler.java]: Both work, documentation changed to reflect properties format

File: src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/config/LinkingManglerRuleConverter.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+import com.thoughtworks.xstream.io.HierarchicalStreamWriter;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import org.geoserver.ows.util.OwsUtils;
+import org.geoserver.proxybase.ext.LinkingManglerRuleBuilder;
+
+/** Converts a {@link LinkingManglerRule} to and from XML. */
+public class LinkingManglerRuleConverter implements Converter {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/config/LinkingManglerRuleConverter.java]: done

File: src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/config/LinkingManglerRuleDAO.java [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+    private static GeoServerDataDirectory getDataDirectory() {
+        return (GeoServerDataDirectory) GeoServerExtensions.bean(DATA_DIRECTORY);
+    }
+
+    /**
+     * Get a temporary path for the {@link LinkingManglerRule} file.
+     *
+     * @return a temporary path for the {@link LinkingManglerRule} file
+     */
+    public static String getTmpLinkingManglerRulesPathPath() {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/config/LinkingManglerRuleDAO.java]: done

File: src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/rest/LinkingManglerRulesController.java [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+@ControllerAdvice
+@RequestMapping(
+        path = LinkingManglerRulesController.RULES_ROOT,
+        produces = {
+            MediaType.APPLICATION_XML_VALUE,
+            MediaType.APPLICATION_JSON_VALUE,
+            MediaTypeExtensions.TEXT_JSON_VALUE
+        })
+/** Controller for the {@link LinkingManglerRule}s. */
+public class LinkingManglerRulesController extends RestBaseController {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/rest/LinkingManglerRulesController.java]: done

File: src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/RuleModel.java [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+ */
+package org.geoserver.proxybase.ext.web;
+
+import java.io.Serializable;
+import java.util.UUID;
+import org.geoserver.proxybase.ext.LinkingManglerRuleBuilder;
+import org.geoserver.proxybase.ext.config.LinkingManglerRule;
+
+/** A model for the {@link LinkingManglerRule}. */
+public class RuleModel implements Serializable {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/RuleModel.java]: removed

File: src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/RulesModel.java [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.geoserver.proxybase.ext.config.LinkingManglerRuleDAO;
+import org.geoserver.web.wicket.GeoServerDataProvider;
+
+/** A model for a collection of the {@link LinkingManglerRule}. */
+public class RulesModel extends GeoServerDataProvider<RuleModel> {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/RulesModel.java]: done

File: src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingConfigPage.java [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+                                if (property == RulesModel.EDIT_BUTTON) {
+                                    return new EditButtonPanel(id, itemModel.getObject());
+                                }
+                                if (property == RulesModel.ACTIVATE_BUTTON) {
+                                    return new ActivateButtonPanel(id, itemModel.getObject());
+                                }
+                                return null;
+                            }
+                        });
+        rulesPanel.setOutputMarkupId(true);
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingConfigPage.java]: done

File: src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingConfigPage.java [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+                                rulesPanel.getSelection().stream()
+                                        .map(RuleModel::getId)
+                                        .toArray(String[]::new));
+                        target.add(rulesPanel);
+                    }
+                });
+        return header;
+    }
+
+    private class EditButtonPanel extends Panel {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingConfigPage.java]: done

File: src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingConfigPage.java [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+                            new AttributeModifier(
+                                    "alt",
+                                    new ParamResourceModel(
+                                            "WebLinkingManglingConfigPage.edit", editLink)));
+            editLink.setOutputMarkupId(true);
+            add(editLink);
+        }
+    }
+
+    private class ActivateButtonPanel extends Panel {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingConfigPage.java]: Checkbox doesn't work as a component, needs to wrapped in a panel.

File: src/community/proxy-base-ext/src/test/java/org/geoserver/proxybase/ext/WebLinkingUrlManglerRestTest.java [hunk 19]
```diff
@@ -0,0 +1,149 @@
...
+import org.junit.Test;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.mock.web.MockHttpServletResponse;
+import org.w3c.dom.Document;
+
+/** Tests the REST API for the web linking mangler */
+public class WebLinkingUrlManglerRestTest extends WebLinkingUrlManglerTestSupport {
+    @Test
+    public void getGetRulesXML() throws Exception {
```
[the remark this answers is not in the document]
Comment (turingtestfail) [reply src/community/proxy-base-ext/src/test/java/org/geoserver/proxybase/ext/WebLinkingUrlManglerRestTest.java]: Validated that auth is required with manual testing, doesn't seem to be required with MockHTTP/GeoServerSystemTestSupport

File: src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingConfigPage.java (hunk 2 above)
Comment (turingtestfail) [reply src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingConfigPage.java]: done

File: src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingRulePage.java (hunk 3 above)
Comment (turingtestfail) [reply src/community/proxy-base-ext/src/main/java/org/geoserver/proxybase/ext/web/WebLinkingManglingRulePage.java]: done

**Pull request merged**